src/main.rs

```diff
@@ -397,17 +332,19 @@ async fn main_impl() -> Result<()> {
                 anyhow::Ok(tmp_path)
             });
 
-            let cancel_token = provider.cancel_token();
+            let provider2 = provider.clone();
             tokio::select! {
                 biased;
                 _ = tokio::signal::ctrl_c() => {
                     println!("Shutting down provider...");
-                    cancel_token.cancel();
+                    provider2.shutdown();
                 }
                 res = provider => {
                     res?;
                 }
             }
+            let snapshot = db.snapshot();
+            snapshot.persist(iroh_data_root)?;
```

Comment:
these blocking calls should probably be wrapped into `spawn_blocking` or `block_in_place`